src/irmin/tree.ml

```diff
@@ -1257,116 +1257,57 @@ module Make (P : Private.S) = struct
     | `Node k -> `Node (Node.of_hash repo k)
     | `Contents (k, m) -> `Contents (Contents.of_hash repo k, m)
 
-  let value_of_map t map = Node.value_of_map t map (fun x -> x)
-
   let export ?clear repo contents_t node_t n =
-    let seen = Hashes.create 127 in
-    let add_node n v () =
-      cnt.node_add <- cnt.node_add + 1;
-      let+ k = P.Node.add node_t v in
-      let k' = Node.hash n in
-      assert (equal_hash k k');
-      Node.export ?clear repo n k
-    in
-    let add_contents c x () =
+    let rec on_node (`Node n) k =
+      match n.Node.v with
+      | Node.Hash (_, h) ->
+          (* Not modified *)
+          Node.export ?clear repo n h;
+          k ()
+      | Node.Value (_, v, None) ->
+          (* Not modified *)
+          let h = P.Node.Key.hash v in
+          Node.export ?clear repo n h;
+          k ()
+      | _ ->
+          let* children_seq =
+            (* TODO: Replace with real sequence getter *)
+            let+ l = Node.list n in
+            get_ok "export" l |> List.to_seq |> Seq.map (fun (_, x) -> x)
+          in
+          on_node_seq children_seq @@ fun () ->
+          cnt.node_add <- cnt.node_add + 1;
+          let* v =
+            (* Turn to inode *)
+            Node.to_value n
+          in
+          let v = get_ok "export" v in
+          let* key = P.Node.add node_t v in
+          let key' = Node.hash n in
+          assert (equal_hash key key');
+          Node.export ?clear repo n key;
+          k ()
+    and on_contents (`Contents (c, _)) k =
       cnt.contents_add <- cnt.contents_add + 1;
-      let+ k = P.Contents.add contents_t x in
-      let k' = Contents.hash c in
-      assert (equal_hash k k');
-      Contents.export ?clear repo c k
-    in
-    let add_node_map n x () = add_node n (value_of_map n x) () in
-    let todo = Stack.create () in
-    let rec add_to_todo : type a. _ -> (unit -> a Lwt.t) -> a Lwt.t =
-     fun n k ->
-      let h = Node.hash n in
-      if Hashes.mem seen h then k ()
```

Comment:
Since the inodes have already been built, this test only prevent the `P.Node.add` part of the work, which would have been a no-op since the hash is already known.